src/dataflow/src/render/context.rs

```diff
@@ -333,40 +338,41 @@ where
         let mode = if key.is_some() { "index" } else { "scan" };
         let name = format!("ArrangementFlatMap({})", mode);
         use timely::dataflow::operators::Operator;
-        trace.stream.unary(Pipeline, &name, move |_, _| {
+        trace.stream.unary(Pipeline, &name, move |_, info| {
+            // Acquire an activator to reschedule the operator when it has unfinished work.
+            use timely::scheduling::Activator;
+            let activations = trace.stream.scope().activations();
+            let activator = Activator::new(&info.address[..], activations);
+            // Maintain a list of work to do, cursor to navigate and process.
+            let mut todo = std::collections::VecDeque::new();
             move |input, output| {
+                // First, dequeue all batches.
                 input.for_each(|time, data| {
-                    let mut session = output.session(&time);
-                    for wrapper in data.iter() {
-                        let batch = &wrapper;
-                        let mut cursor = batch.cursor();
-                        if let Some(key) = &key {
-                            cursor.seek_key(batch, key);
-                            if cursor.get_key(batch) == Some(key) {
-                                while let Some(val) = cursor.get_val(batch) {
-                                    cursor.map_times(batch, |time, diff| {
-                                        for datum in logic(RefOrMut::Ref(val), time, diff) {
-                                            session.give(datum);
-                                        }
-                                    });
-                                    cursor.step_val(batch);
-                                }
-                            }
-                        } else {
-                            while let Some(_key) = cursor.get_key(batch) {
-                                while let Some(val) = cursor.get_val(batch) {
-                                    cursor.map_times(batch, |time, diff| {
-                                        for datum in logic(RefOrMut::Ref(val), time, diff) {
-                                            session.give(datum);
-                                        }
-                                    });
-                                    cursor.step_val(batch);
-                                }
-                                cursor.step_key(batch);
-                            }
-                        }
+                    let capability = time.retain();
+                    for batch in data.iter() {
+                        // enqueue a capability, cursor, and batch.
+                        todo.push_back(PendingWork::new(
+                            capability.clone(),
+                            batch.cursor(),
+                            batch.clone(),
```

Comment:
What cost do we think the copy of the batch incurs here? Would it be possible to leave batches in the input if we're out of fuel instead?

Reply:
It is intended to be very cheap (it is a `Rc` just under covers). I think it's either documented or otherwise intended that clones should be cheap, as that is the form of sharing between different users. I think leaving it in the input is un-idiomatic, in that it blocks other workers from knowing that all batches have been consumed, which might be blocking other actions on their part (probably not, because we can see the operator doesn't care, but "unidiomatic"). I think peeling it out of the input and stashing it somewhere, roughly like this, ... well that is what I think is idiomatic, but I could be convinced otherwise.

Reply:
Thanks for the explanation! The cost of cloning an `Rc` should be acceptable here.